lens/fetcher.go

```diff
@@ -0,0 +1,400 @@
+		if !fieldFound {
+			// Note: This can technically happen if a Lens migration returns a field that
+			// we do not know about. In which case we have to skip it.
+			continue
+		}
+
+		fieldValue, err := core.DecodeFieldValue(fieldDesc, fieldByteValue)
+		if err != nil {
+			return nil, err
+		}
+
+		err = doc.SetAs(fieldDesc.Name, fieldValue, fieldDesc.Typ)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	doc.SchemaVersionID = f.col.Schema.VersionID
+
+	// Note: client.Document does not have a means of flagging as to whether it is
+	// deleted or not, and, currently the fetcher does not ever returned deleted items
+	// from the function that returs this type.
+
+	return doc, nil
+}
+
+// updateDataStore updates the datastore with the migrated values.
+//
+// This removes the need to migrate a document everytime it is fetched as the second time around
+// the underlying fetcher will return the migrated values cached in the datastore instead of the
+// underlying dag store values.
+func (f *lensedFetcher) updateDataStore(ctx context.Context, original map[string]any, migrated map[string]any) error {
+	modifiedFieldValuesByName := map[string]any{}
+	for name, originalValue := range original {
+		migratedValue, ok := migrated[name]
+		if !ok {
+			// If the field is present in the original, and missing from the migrated, it
+			// means that a migration has removed it, and we should set it to nil.
+			modifiedFieldValuesByName[name] = nil
+			continue
+		}
+
+		// Note: A deep equals check is required here, as the values may be inline-array slices
+		// Todo: `reflect.DeepEqual` is pretty rubish long-term here and should be replaced
+		// with something more defra specific: https://github.com/sourcenetwork/defradb/issues/1606
+		if !reflect.DeepEqual(originalValue, migratedValue) {
+			modifiedFieldValuesByName[name] = migratedValue
+		}
+	}
+
+	for name, migratedValue := range migrated {
+		if _, ok := original[name]; !ok {
+			// If a field has been added by a migration we need to make sure we
+			// preserve it here.
+			modifiedFieldValuesByName[name] = migratedValue
+			continue
+		}
+	}
+
+	if len(modifiedFieldValuesByName) > 0 {
+		dockey, ok := original[request.KeyFieldName].(string)
+		if !ok {
+			return core.ErrInvalidKey
+		}
+
+		datastoreKeyBase := core.DataStoreKey{
+			CollectionID: f.col.IDString(),
+			DocKey:       dockey,
+			InstanceType: core.ValueKey,
+		}
+
+		for fieldName, value := range modifiedFieldValuesByName {
+			fieldDesc, ok := f.fieldDescriptionsByName[fieldName]
+			if !ok {
+				// It may be that the migration has set fields that are unknown to us locally
+				// in which case we have to skip them for now.
+				continue
+			}
+			fieldKey := datastoreKeyBase.WithFieldId(fieldDesc.ID.String())
+
+			bytes, err := cbor.Marshal(value)
+			if err != nil {
+				return err
+			}
+
+			err = f.txn.Datastore().Put(ctx, fieldKey.ToDS(), append([]byte{byte(fieldDesc.Typ)}, bytes...))
```

Comment:
note: one more place with the knowledge of the CRDT byte. We should definitely refactor it soon